homeassistant/components/media_player/songpal.py

```diff
@@ -0,0 +1,252 @@
+    def unique_id(self):
+        """Return an unique ID."""
+        return self._sysinfo.macAddr
+
+    @property
+    def available(self):
+        """Return availability of the device."""
+        return self._available
+
+    async def async_set_sound_setting(self, name, value):
+        """Change a setting on the device."""
+        await self.dev.set_sound_settings(name, value)
+
+    async def async_update(self):
+        """Fetch updates from the device."""
+        from songpal import SongpalException
+        try:
+            volumes = await self.dev.get_volume_information()
+            if not volumes:
+                _LOGGER.error("Got no volume controls, bailing out")
+                self._available = False
+                return
+
+            if len(volumes) > 1:
+                _LOGGER.warning("Got %s volume controls, using the first one",
+                                volumes)
+
+            volume = volumes.pop()
+            _LOGGER.debug("Current volume: %s", volume)
+
+            self._volume_max = volume.maxVolume
+            self._volume_min = volume.minVolume
+            self._volume = volume.volume
+            self._volume_control = volume
+            self._is_muted = self._volume_control.is_muted
+
+            status = await self.dev.get_power()
+            self._state = status.status
+            _LOGGER.debug("Got state: %s", status)
+
+            inputs = await self.dev.get_inputs()
+            _LOGGER.debug("Got ins: %s", inputs)
+            self._sources = inputs
+
+            self._available = True
+        except SongpalException as ex:
+            # if we were available, print out the exception
+            if self._available:
+                _LOGGER.error("Got an exception: %s", ex)
+            self._available = False
+
+    async def async_select_source(self, source):
+        """Select source."""
+        for out in self._sources:
+            if out.title == source:
+                await out.activate()
+                return
+
+        _LOGGER.error("Unable to find output: %s", source)
+
+    @property
+    def source_list(self):
+        """Return list of available sources."""
+        return [x.title for x in self._sources]
+
+    @property
+    def state(self):
+        """Return current state."""
+        if self._state:
+            return STATE_ON
+        return STATE_OFF
+
+    @property
+    def source(self):
+        """Return currently active source."""
+        for out in self._sources:
+            if out.active:
+                return out.title
+
+        return None
+
+    @property
+    def volume_level(self):
+        """Return volume level."""
+        volume = self._volume / self._volume_max
+        return volume
+
+    async def async_set_volume_level(self, volume):
+        """Set volume level."""
+        volume = int(volume * self._volume_max)
+        _LOGGER.debug("Setting volume to %s", volume)
+        return await self._volume_control.set_volume(volume)
```

Comment:
Why return here?

Reply:
That is indeed unnecessary here, but it should not cause any side effects so I will fix that with the same PR where this will be converted non-polling, if that's okay.

Reply:
:+1: